runtime/lua/vim/fs.lua

```diff
@@ -349,22 +349,24 @@ function M.normalize(path, opts)
   })
 
   if path:sub(1, 1) == '~' then
-    local home = vim.uv.os_homedir() or '~'
-    if home:sub(-1) == '\\' or home:sub(-1) == '/' then
-      home = home:sub(1, -2)
-    end
-    path = home .. path:sub(2)
+    path = (vim.uv.os_homedir() or '~') .. '/' .. path:sub(2)
   end
 
   if opts.expand_env == nil or opts.expand_env then
     path = path:gsub('%$([%w_]+)', vim.uv.os_getenv)
   end
 
-  path = path:gsub('\\', '/'):gsub('/+', '/')
-  if iswin and path:match('^%w:/$') then
-    return path
+  local first = path:sub(1, 1)
+  path = (first == '\\' and '/' or first) .. path:sub(2):gsub('[\\/]+', '/'):gsub('/$', '')
+  -- UNC path starts with double (or more) slash, followed by more than one character.
+  -- Therefore ignore first letter from `:gsub('[\\/]+', '/')`.
+  -- `//` is not a valid UNC path, hence converted to `/` with `:gsub('/$', '')`.
+  -- This also keeps traling slash in the case of root `/`, but removes it otherwise.
+
+  if iswin and path:sub(2, 2) == ':' then -- capitalize windows drive name
```

Comment:
Should probably check the first char is `[a-zA-Z]`

Reply:
I don't think so. Colon is not a valid character for a path string except right after the device name.

And even if it is true, I'm just doing `string.upper(first)` which does nothing when first is not an alphabet in the first place.

Reply:
This function needs to gracefully handle any old junk that's passed to it, so it needs to be sensible with invalid paths. So we should only mangle if the path matches `^[a-zA-Z]:/`, otherwise we just leave it alone.

Reply:
I mean that `string.upper` will do nothing if the first letter is not an alphabet. Am I right?

Reply:
@lewis6991 Do I need to change anything here?

Reply:
Yes, this code assumes `path[3] == '/'` which it might not be. I think we need to match `^[a-zA-Z]:/` exactly here so we only mangle if we absolutely need to.

Reply:
No. Due to `path:sub(2):...:gsub('/$', '')`, `C:/` will be converted to `C:` temporary and a slash must be added afterwards to negate this behavior.

Please read the code.

Reply:
> Please read the code.

Please be polite. I don't have to review this.

> Due to path:sub(2):...:gsub('/$', ''), C:/ will be converted to C: temporary and a slash must be added afterwards to negate this behavior.

This is not obvious at all. And despite this, it still assumes `path[3] == '/'` for cases other than `C:/`

Reply:
I thought you understood this while writing https://github.com/neovim/neovim/pull/27202#discussion_r1474262585? 
It is the exact behavior you wrote in the comments.

> And despite this, it still assumes `path[3] == '/'` for cases other than `C:/`

In [this issue](https://github.com/neovim/neovim/issues/27068) you refused to bring UNC paths because it is not supported by the Unix path standard. And now I feel like you are trying to support all edge cases for every file system. May I ask what direction you are towards right now?

Besides that, is there an edge case where `iswin and path:sub(2, 2) == ':'` but `path[3] ~= '/'`?

Reply:
> I thought you understood this while writing https://github.com/neovim/neovin/pull/27202#discussion_r1474262585?
It is the exact behavior you wrote in the comments.

No, not at all. That specific code is all about normalizing slashes AND preserving leading `//`. It has nothing to do with the single `C:/` edge case.

> In https://github.com/neovim/neovim/issues/27068 you refused to bring UNC paths because it is not supported by the Unix path standard. And now I feel like you are trying to support all edge cases for every file system. May I ask what direction you are towards right now?

We support it or we don't support it. We've decided to support it and there is a glaring bug in this code.

To put it simply.

```
C:some_invalid_path
```
should NOT normalize to:
```
C:/ome_invalid_path
```

Reply:
Thank you for your input. I find your words quite hurtful, so I would prefer not to contribute anymore.

I wish the best of luck for this project.